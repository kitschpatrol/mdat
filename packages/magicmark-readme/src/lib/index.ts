export {
	type CheckReadmeFileOptions,
	type CheckReadmeFileReport,
	type CheckReadmeStringOptions,
	type CheckReadmeStringReport,
	checkReadmeFile,
	checkReadmeString,
} from './check'

export {
	type ExpandReadmeFileOptions,
	type ExpandReadmeFileReport,
	type ExpandReadmeStringOptions,
	type ExpandReadmeStringReport,
	expandReadmeFile,
	expandReadmeString,
} from './expand'
