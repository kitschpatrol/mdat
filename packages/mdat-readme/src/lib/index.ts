export {
	type ExpandReadmeFileReport,
	type ExpandReadmeOptions,
	type ExpandReadmeStringReport,
	expandReadmeFile,
	expandReadmeString,
} from './api'
