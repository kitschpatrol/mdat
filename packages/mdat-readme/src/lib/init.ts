// This will do something soon
console.log('Hello, world!')
