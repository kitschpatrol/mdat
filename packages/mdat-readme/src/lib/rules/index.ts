import badges from './badges'
import banner from './banner'
import cliHelp from './cli-help'
import code from './code'
import contributing from './contributing'
import footer from './footer'
import header from './header'
import license from './license'
import shortDescription from './short-description'
import tableOfContents from './table-of-contents'
import title from './title'
import toc from './toc'
import { type Rules } from 'remark-mdat'

// A little redundant to have an expander name / preset in both
// the record keys and the `keyword` field in the expansion object,
// but this smooths over the kebab vs. camel case difference and arguably
// makes expansions more portable

export default {
	...badges,
	...banner,
	...cliHelp,
	...code,
	...contributing,
	...footer,
	...header,
	...license,
	...shortDescription,
	...tableOfContents,
	...toc,
	...title,
} satisfies Rules
