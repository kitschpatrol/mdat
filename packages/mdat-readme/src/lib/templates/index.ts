// Bundle template markdown for convenience

// Note:
// This project is build with tsup, not vite... but we've configured a custom
// loader for .md files in tsup.config.ts and a module declaration in
// ./src/types/md.d.ts. We use the ?raw suffix to intersect with the vite loader
// and get the raw file contents in the vitest context, without actually having
// to include vite as a direct dev dependency.

import readmeBasicCompound from './readme-basic-compound.md?raw'
import readmeBasic from './readme-basic.md?raw'
import readmeFullCompound from './readme-full-compound.md?raw'
import readmeFull from './readme-full.md?raw'

export default {
	readmeBasic,
	readmeBasicCompound,
	readmeFull,
	readmeFullCompound,
}
