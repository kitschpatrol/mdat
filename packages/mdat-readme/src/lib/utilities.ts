import { findUp } from 'find-up'
import { type NormalizedRules, type Rules, log } from 'mdat'
import path from 'node:path'
import { packageUp } from 'package-up'
import { packageDirectory } from 'pkg-dir'
import { type NormalizedPackageJson, readPackage } from 'read-pkg'

/**
 * Searches for a readme file in the following order:
 * 1. Searches the current working directly for readme.md
 * 2. If there's no readme.md in the current directory, search up to the closest package directory
 * 3. Give up and return undefined if no readme is found
 *
 * @returns The path to the readme file
 * @throws If no readme is found
 */
export async function findReadme(): Promise<string | undefined> {
	log.info(`Searching for package directory...`)
	// Treat the closest package directory, if available, as the "find up" limit
	const searchCeilingDirectory = (await packageDirectory()) ?? process.cwd()

	// FindUp is case-insensitive
	const closestReadme = await findUp('readme.md', {
		stopAt: searchCeilingDirectory,
		type: 'file',
	})

	if (closestReadme !== undefined) {
		log.info(`Found closest readme at "${closestReadme}"`)
		return closestReadme
	}

	throw new Error('No readme found')
}

// Decadent
// function getCasePermutations(input: string, index = 0, current = ''): string[] {
// 	if (index === input.length) {
// 		return [current]
// 	}

// 	const char = input[index]
// 	const nextPermutations = getCasePermutations(input, index + 1, current + char)

// 	if (char.toLowerCase() !== char.toUpperCase()) {
// 		const upperPermutations = getCasePermutations(input, index + 1, current + char.toUpperCase())
// 		return [...nextPermutations, ...upperPermutations]
// 	}

// 	return nextPermutations
// }

export async function findPackage(): Promise<string> {
	const packageFile = await packageUp()
	if (packageFile === undefined) {
		throw new Error('No package.json found')
	}

	return packageFile
}

// Careful, global stuff here...
// For use by rules, allows override of package.json
// via cli options
let packageJson: NormalizedPackageJson | undefined
let packageFile: string | undefined

export function setPackageFile(path: string) {
	packageFile = path
}

// Load as package json only as needed, memoize
export async function getPackageJson(): Promise<NormalizedPackageJson> {
	// Find package as needed
	packageFile ??= await findPackage()
	packageJson ??= await readPackage({ cwd: path.dirname(packageFile) })

	if (packageJson === undefined) {
		throw new Error('No package.json found')
	}

	return packageJson
}

// Sometimes rules need the path to the package.json
export async function getPackagePath(): Promise<string> {
	// Find package as needed
	packageFile ??= await findPackage()
	return packageFile
}

export function getSoleRule<T extends NormalizedRules | Rules>(rules: T): T[keyof T] {
	return getSoleRecord<T[keyof T]>(rules as Record<string, T[keyof T]>)
}

export function getSoleRuleKey<T extends NormalizedRules | Rules>(rules: T): keyof T {
	const keys = Object.keys(rules)
	if (keys.length !== 1) {
		throw new Error(`Expected exactly one rule, found ${keys.length}`)
	}

	return keys[0]
}

/**
 * Get the sole entry in a record.
 *
 * Useful for working with Rules records
 * that are only supposed to contain a single rule.
 *
 * @param record The record to get the sole entry from
 * @returns The value of the sole entry in the record
 * @throws If there are no entries or more than one entry
 */
export function getSoleRecord<V>(record: Record<string, V>): V {
	const recordValues = Object.values(record)
	if (recordValues.length === 0) {
		throw new Error('Found no entries in a "sole record" record. This should never happen')
	}

	if (recordValues.length > 1) {
		throw new Error('Found multiple entries in "sole record" record. This should never happen')
	}

	return recordValues[0]
}
