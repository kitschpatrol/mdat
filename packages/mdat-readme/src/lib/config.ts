import readmeRules from './rules'
import { findPackage, findReadme } from './utilities'
import { type Options as MdatConfig } from 'mdat'
import { loadConfig } from 'mdat'
import { z } from 'zod'

export type MdatReadmeConfig = {
	packageFile?: string
	readmeFile?: string
} & MdatConfig

const mdatReadmeConfigExtensionSchema = z
	.object({
		packageFile: z.string().optional(),
		readmeFile: z.string().optional(),
	})
	.describe('Mdat Readme Config Extension')

async function getReadmeDefaults(): Promise<MdatReadmeConfig> {
	return {
		addMetaComment: true,
		packageFile: await findPackage(),
		readmeFile: await findReadme(),
		rules: readmeRules,
	}
}

// Convenience loader to always include the default readme config
// Note passing config type to loadConfig parameter plucker
type LoadConfigOptions = Parameters<typeof loadConfig<MdatReadmeConfig>>[0]
export async function loadConfigReadme(options?: LoadConfigOptions): Promise<MdatReadmeConfig> {
	const { additionalConfigsOrRules = [], configExtensionSchema, searchFrom } = options ?? {}
	const readmeDefaultConfig = await getReadmeDefaults()

	const result = await loadConfig<MdatReadmeConfig>({
		additionalConfigsOrRules: [readmeDefaultConfig, ...additionalConfigsOrRules],
		configExtensionSchema:
			configExtensionSchema === undefined
				? mdatReadmeConfigExtensionSchema
				: mdatReadmeConfigExtensionSchema.merge(configExtensionSchema),
		searchFrom, // Unchanged
	})

	return result
}
