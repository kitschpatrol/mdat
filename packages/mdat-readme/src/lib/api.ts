import { type MdatReadmeConfig, loadConfigReadme } from './config'
import { setPackageFile } from './utilities'
import { type ExpandOptions, expandFile, expandString } from 'mdat'
import type { Simplify } from 'type-fest'
import { type VFile } from 'vfile'

export type ExpandReadmeOptions = ExpandOptions<MdatReadmeConfig>

// Expand a readme file, maps to CLI
export type ExpandReadmeFileReport = Simplify<
	ExpandReadmeStringReport & {
		readmeFile: string
	}
>

export async function expandReadmeFile(
	options?: ExpandReadmeOptions,
): Promise<ExpandReadmeFileReport> {
	const resolvedOptions = await loadConfigReadme({ additionalConfigsOrRules: options })
	const { packageFile, readmeFile, ...expandFileOptions } = resolvedOptions

	// This should never happen because the defaults are set in loadConfigReadme
	if (packageFile === undefined || readmeFile === undefined) {
		throw new Error('Package and readme files are required')
	}

	// Hacky global state for rules
	setPackageFile(packageFile)

	// No second deep load because expandFileOptions already has loaded and normalized rules
	const result = await expandFile(readmeFile, undefined, undefined, [expandFileOptions])

	return {
		packageFile,
		readmeFile,
		result,
	}
}

// Expand a readme string, mostly for testing
export type ExpandReadmeStringReport = {
	packageFile: string
	result: VFile
}

export async function expandReadmeString(
	markdown: string,
	options?: ExpandReadmeOptions,
): Promise<ExpandReadmeStringReport> {
	const resolvedOptions = await loadConfigReadme({ additionalConfigsOrRules: options })
	const { packageFile, readmeFile, ...expandOptions } = resolvedOptions

	// This should never happen because the defaults are set
	if (packageFile === undefined || readmeFile === undefined) {
		throw new Error('Package and readme files are required')
	}

	// Hacky global state for rules
	setPackageFile(packageFile)

	const result = await expandString(markdown, [expandOptions])

	return {
		packageFile,
		result,
	}
}
