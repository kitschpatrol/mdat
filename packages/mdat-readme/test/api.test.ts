import { expandReadmeFile, expandReadmeString } from '../src/lib/api'
import fs from 'node:fs/promises'
import { describe, expect, it } from 'vitest'

// Replace matched dates with the placeholder text for stable snapshots
function stripDynamic(text: string): string {
	return text.replaceAll(/\s\d{4}-\d{2}-\d{2}\s/g, ' ****-**-** ')
}

describe('comment expansion', () => {
	it('should expand readme comments', async () => {
		const markdown = await fs.readFile('./test/assets/readme-test.md', 'utf8')
		const { result } = await expandReadmeString(markdown)
		expect(stripDynamic(result.toString())).toMatchSnapshot()
	})

	it('should expand prefixed comments only', async () => {
		const markdown = await fs.readFile('./test/assets/readme-test.md', 'utf8')
		const { result } = await expandReadmeString(markdown, {
			keywordPrefix: 'mm-',
		})

		expect(stripDynamic(result.toString())).toMatchSnapshot()
	})

	it('should allow additional rules, and they should override those provided my mdat-readme', async () => {
		const markdown = await fs.readFile('./test/assets/readme-test.md', 'utf8')
		const { result } = await expandReadmeString(markdown, undefined, './test/assets/extra-rules.js')
		expect(stripDynamic(result.toString())).toMatchSnapshot()
	})

	it('should exclude the meta tag if asked', async () => {
		const markdown = await fs.readFile('./test/assets/readme-test.md', 'utf8')
		const { result } = await expandReadmeString(markdown, {
			addMetaComment: false,
		})
		expect(result.toString()).toMatchSnapshot()
	})

	it('should find the local readme and package.json and expand them', async () => {
		const { packageFile, readmeFile, result } = await expandReadmeFile()
		expect(packageFile).not.toBeUndefined()
		expect(readmeFile).not.toBeUndefined()
		expect(result.toString()).toContain('<!-- /')
	}, 30_000)
})
