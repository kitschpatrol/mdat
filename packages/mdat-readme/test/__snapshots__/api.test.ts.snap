// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`comment expansion > should allow additional rules, and they should override those provided my mdat-readme 1`] = `
VFile {
  "cwd": "/Users/mika/Code/mdat/packages/mdat-readme",
  "data": {},
  "history": [],
  "messages": [
    [3:1-3:1: Expanded: <!-- banner -->],
    [5:1-5:1: Expanded: <!-- badges {custom: {Documentation: {image: "https://img.shields.io/badge/-documentation-ffdd00?logo=readthedocs&logoColor=222222", link: "https://kitschpatrol.com/svelte-tweakpane-ui"}}} -->],
    [7:1-7:1: Expanded: <!-- short-description -->],
    [17:1-17:1: Expanded: <!-- code {file: "./test/assets/extra-rules.js"} -->],
    [29:1-29:1: Expanded: <!-- extra -->],
    [31:1-31:1: Expanded: <!-- contributing -->],
    [33:1-33:1: Expanded: <!-- license -->],
    [9:1-9:1: Expanded: <!-- table-of-contents -->],
    [1:1-1:1: Expanded: <!-- title {prefix: 'üìù '} -->],
    [1:1-1:1: Missing optional: <!-- footer -->],
    [1:1-1:1: Missing optional: <!-- header -->],
    [23:1-23:1: Missing rule: <!-- mm-short-description  -->],
  ],
  "value": "<!--+ Warning: Content in HTML comment blocks generated by mdat on 2024-02-01 +-->

<!-- title {prefix: 'üìù '} -->

# \`extra-rules.js\` has overridden the title

<!-- /title -->

<!-- banner -->

![mdat-readme banner](assets/hero.png)

<!-- /banner -->

<!-- badges {custom: {Documentation: {image: "https://img.shields.io/badge/-documentation-ffdd00?logo=readthedocs&logoColor=222222", link: "https://kitschpatrol.com/svelte-tweakpane-ui"}}} -->

[![NPM Package](https://img.shields.io/npm/v/mdat-readme.svg)](https://npmjs.com/package/mdat-readme)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)
[![Documentation](https://img.shields.io/badge/-documentation-ffdd00?logo=readthedocs\\&logoColor=222222)](https://kitschpatrol.com/svelte-tweakpane-ui)

<!-- /badges -->

<!-- short-description -->

**Automatically populate comment placeholders in readme.md files with metadata from your package.json.**

<!-- /short-description -->

<!-- table-of-contents -->

## Table of contents

* [Example code](#example-code)

* [Prefix example](#prefix-example)

* [Extra rules example](#extra-rules-example)

* [Contributing](#contributing)

* [License](#license)

<!-- /table-of-contents -->

A long description goes here.

## Example code

The code block below was dynamically embedded by Mdat.

<!-- code {file: "./test/assets/extra-rules.js"} -->

\`\`\`.js
// This is a code sample for testing purposes
// There doesn't seem to be a way to type a default export in JSDoc

/** @satisfies {import('../../src/lib/rules').Rules} */
const rules = {
	extra: '**A bold statement from extra-rules.js**',
	title: {
		applicationOrder: 2,
		content: '# \`extra-rules.js\` has overridden the title',
	},
}

export default rules

\`\`\`

<!-- /code -->

## Prefix example

The comment below only expands if \`--prefix mm-\` was set:

<!-- mm-short-description  -->

## Extra rules example

The comment below only expands if \`--rules extra-rules.js\` was set:

<!-- extra -->

**A bold statement from extra-rules.js**

<!-- /extra -->

<!-- contributing -->

## Contributing

[Issues](https://github.com/kitschpatrol/mdat/issues) and pull requests are welcome.

<!-- /contributing -->

<!-- license -->

## License

[MIT](license.txt) ¬© Eric Mika

<!-- /license -->
",
}
`;

exports[`comment expansion > should exclude the meta tag if asked 1`] = `
VFile {
  "cwd": "/Users/mika/Code/mdat/packages/mdat-readme",
  "data": {},
  "history": [],
  "messages": [
    [3:1-3:1: Expanded: <!-- banner -->],
    [5:1-5:1: Expanded: <!-- badges {custom: {Documentation: {image: "https://img.shields.io/badge/-documentation-ffdd00?logo=readthedocs&logoColor=222222", link: "https://kitschpatrol.com/svelte-tweakpane-ui"}}} -->],
    [7:1-7:1: Expanded: <!-- short-description -->],
    [17:1-17:1: Expanded: <!-- code {file: "./test/assets/extra-rules.js"} -->],
    [31:1-31:1: Expanded: <!-- contributing -->],
    [33:1-33:1: Expanded: <!-- license -->],
    [9:1-9:1: Expanded: <!-- table-of-contents -->],
    [1:1-1:1: Expanded: <!-- title {prefix: 'üìù '} -->],
    [1:1-1:1: Missing optional: <!-- footer -->],
    [1:1-1:1: Missing optional: <!-- header -->],
    [23:1-23:1: Missing rule: <!-- mm-short-description  -->],
    [29:1-29:1: Missing rule: <!-- extra -->],
  ],
  "value": "<!-- title {prefix: 'üìù '} -->

# üìù mdat-readme

<!-- /title -->

<!-- banner -->

![mdat-readme banner](assets/hero.png)

<!-- /banner -->

<!-- badges {custom: {Documentation: {image: "https://img.shields.io/badge/-documentation-ffdd00?logo=readthedocs&logoColor=222222", link: "https://kitschpatrol.com/svelte-tweakpane-ui"}}} -->

[![NPM Package](https://img.shields.io/npm/v/mdat-readme.svg)](https://npmjs.com/package/mdat-readme)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)
[![Documentation](https://img.shields.io/badge/-documentation-ffdd00?logo=readthedocs\\&logoColor=222222)](https://kitschpatrol.com/svelte-tweakpane-ui)

<!-- /badges -->

<!-- short-description -->

**Automatically populate comment placeholders in readme.md files with metadata from your package.json.**

<!-- /short-description -->

<!-- table-of-contents -->

## Table of contents

* [Example code](#example-code)

* [Prefix example](#prefix-example)

* [Extra rules example](#extra-rules-example)

* [Contributing](#contributing)

* [License](#license)

<!-- /table-of-contents -->

A long description goes here.

## Example code

The code block below was dynamically embedded by Mdat.

<!-- code {file: "./test/assets/extra-rules.js"} -->

\`\`\`.js
// This is a code sample for testing purposes
// There doesn't seem to be a way to type a default export in JSDoc

/** @satisfies {import('../../src/lib/rules').Rules} */
const rules = {
	extra: '**A bold statement from extra-rules.js**',
	title: {
		applicationOrder: 2,
		content: '# \`extra-rules.js\` has overridden the title',
	},
}

export default rules

\`\`\`

<!-- /code -->

## Prefix example

The comment below only expands if \`--prefix mm-\` was set:

<!-- mm-short-description  -->

## Extra rules example

The comment below only expands if \`--rules extra-rules.js\` was set:

<!-- extra -->

<!-- contributing -->

## Contributing

[Issues](https://github.com/kitschpatrol/mdat/issues) and pull requests are welcome.

<!-- /contributing -->

<!-- license -->

## License

[MIT](license.txt) ¬© Eric Mika

<!-- /license -->
",
}
`;

exports[`comment expansion > should expand prefixed comments only 1`] = `
VFile {
  "cwd": "/Users/mika/Code/mdat/packages/mdat-readme",
  "data": {},
  "history": [],
  "messages": [
    [23:1-23:1: Expanded: <!-- mm-short-description  -->],
    [1:1-1:1: Missing required: <!-- contributing -->],
    [1:1-1:1: Missing required: <!-- license -->],
    [1:1-1:1: Missing required: <!-- title -->],
    [1:1-1:1: Missing optional: <!-- badges -->],
    [1:1-1:1: Missing optional: <!-- banner -->],
    [1:1-1:1: Missing optional: <!-- code -->],
    [1:1-1:1: Missing optional: <!-- footer -->],
    [1:1-1:1: Missing optional: <!-- header -->],
    [1:1-1:1: Missing optional: <!-- table-of-contents -->],
    [1:1-1:1: Missing prefix: <!-- title {prefix: 'üìù '} -->],
    [5:1-5:1: Missing prefix: <!-- badges {custom: {Documentation: {image: "https://img.shields.io/badge/-documentation-ffdd00?logo=readthedocs&logoColor=222222", link: "https://kitschpatrol.com/svelte-tweakpane-ui"}}} -->],
    [17:1-17:1: Missing prefix: <!-- code {file: "./test/assets/extra-rules.js"} -->],
    [29:1-29:1: Missing prefix: <!-- extra -->],
  ],
  "value": "<!--+ Warning: Content in HTML comment blocks generated by mdat on 2024-02-01 +-->

<!-- title {prefix: 'üìù '} -->

<!-- banner -->

<!-- badges {custom: {Documentation: {image: "https://img.shields.io/badge/-documentation-ffdd00?logo=readthedocs&logoColor=222222", link: "https://kitschpatrol.com/svelte-tweakpane-ui"}}} -->

<!-- short-description -->

<!-- table-of-contents -->

A long description goes here.

## Example code

The code block below was dynamically embedded by Mdat.

<!-- code {file: "./test/assets/extra-rules.js"} -->

## Prefix example

The comment below only expands if \`--prefix mm-\` was set:

<!-- mm-short-description  -->

**Automatically populate comment placeholders in readme.md files with metadata from your package.json.**

<!-- /mm-short-description -->

## Extra rules example

The comment below only expands if \`--rules extra-rules.js\` was set:

<!-- extra -->

<!-- contributing -->

<!-- license -->
",
}
`;

exports[`comment expansion > should expand readme comments 1`] = `
VFile {
  "cwd": "/Users/mika/Code/mdat/packages/mdat-readme",
  "data": {},
  "history": [],
  "messages": [
    [3:1-3:1: Expanded: <!-- banner -->],
    [5:1-5:1: Expanded: <!-- badges {custom: {Documentation: {image: "https://img.shields.io/badge/-documentation-ffdd00?logo=readthedocs&logoColor=222222", link: "https://kitschpatrol.com/svelte-tweakpane-ui"}}} -->],
    [7:1-7:1: Expanded: <!-- short-description -->],
    [17:1-17:1: Expanded: <!-- code {file: "./test/assets/extra-rules.js"} -->],
    [31:1-31:1: Expanded: <!-- contributing -->],
    [33:1-33:1: Expanded: <!-- license -->],
    [9:1-9:1: Expanded: <!-- table-of-contents -->],
    [1:1-1:1: Expanded: <!-- title {prefix: 'üìù '} -->],
    [1:1-1:1: Missing optional: <!-- footer -->],
    [1:1-1:1: Missing optional: <!-- header -->],
    [23:1-23:1: Missing rule: <!-- mm-short-description  -->],
    [29:1-29:1: Missing rule: <!-- extra -->],
  ],
  "value": "<!--+ Warning: Content in HTML comment blocks generated by mdat on 2024-02-01 +-->

<!-- title {prefix: 'üìù '} -->

# üìù mdat-readme

<!-- /title -->

<!-- banner -->

![mdat-readme banner](assets/hero.png)

<!-- /banner -->

<!-- badges {custom: {Documentation: {image: "https://img.shields.io/badge/-documentation-ffdd00?logo=readthedocs&logoColor=222222", link: "https://kitschpatrol.com/svelte-tweakpane-ui"}}} -->

[![NPM Package](https://img.shields.io/npm/v/mdat-readme.svg)](https://npmjs.com/package/mdat-readme)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)
[![Documentation](https://img.shields.io/badge/-documentation-ffdd00?logo=readthedocs\\&logoColor=222222)](https://kitschpatrol.com/svelte-tweakpane-ui)

<!-- /badges -->

<!-- short-description -->

**Automatically populate comment placeholders in readme.md files with metadata from your package.json.**

<!-- /short-description -->

<!-- table-of-contents -->

## Table of contents

* [Example code](#example-code)

* [Prefix example](#prefix-example)

* [Extra rules example](#extra-rules-example)

* [Contributing](#contributing)

* [License](#license)

<!-- /table-of-contents -->

A long description goes here.

## Example code

The code block below was dynamically embedded by Mdat.

<!-- code {file: "./test/assets/extra-rules.js"} -->

\`\`\`.js
// This is a code sample for testing purposes
// There doesn't seem to be a way to type a default export in JSDoc

/** @satisfies {import('../../src/lib/rules').Rules} */
const rules = {
	extra: '**A bold statement from extra-rules.js**',
	title: {
		applicationOrder: 2,
		content: '# \`extra-rules.js\` has overridden the title',
	},
}

export default rules

\`\`\`

<!-- /code -->

## Prefix example

The comment below only expands if \`--prefix mm-\` was set:

<!-- mm-short-description  -->

## Extra rules example

The comment below only expands if \`--rules extra-rules.js\` was set:

<!-- extra -->

<!-- contributing -->

## Contributing

[Issues](https://github.com/kitschpatrol/mdat/issues) and pull requests are welcome.

<!-- /contributing -->

<!-- license -->

## License

[MIT](license.txt) ¬© Eric Mika

<!-- /license -->
",
}
`;
